### PR TITLE
Display constant sum statistics sorted by recipient name or option #3147

### LINKED ISSUE #3147: Display constant sum statistics sorted by recipient name or option
For both the statistics on the website and in the csv file downloadable by instructors, the statistics table can be displayed in a sorted order

Edit (18 December 2016):
Screenshot (of the table to change sorting for ): 
constantsumstats

### PR DESCRIPTION
**Pull request by swxsw**
Fixes #3147
Update FeedbackConstantSumQuestionDetails to sort by option or name,
using SortedMap and TreeMap

### CODE REVIEW

**Comment by swxsw** [comment]
Writing tests, in progress

**Comment by swxsw** [comment]
Should be ready for review, a little unsure about commit `b1da49c`

**Comment by damithc** [comment]
@swxsw while you wait for a review, may be you can syn the branch with upstream master?

**Comment by JosephineKwa** [comment]
Hi @swxsw, do change the PR title according to the format specified [here](https://github.com/TEAMMATES/teammates/blob/master/docs/process.md#creating-a-pull-request). The issue number should be at the end of the title.

**Comment by swxsw** [comment]
Done:)

**Comment by damithc** [comment]
For future reference, branch name should be *some* keywords from the issue title, not *all* words from the issue title.

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 1]
```diff
@@ -398,7 +400,16 @@ public String getQuestionResultStatisticsHtml(
...
         
-        for (Entry<String, List<Integer>> entry : optionPoints.entrySet()) {
+        SortedMap<String, List<Integer>> sortedOptionPoints = new TreeMap<String, List<Integer>>(new Comparator<String>() {
+            @Override
+            public int compare(String s1, String s2) {
+                return s2.compareTo(s1);
+            }
+        });
+        
+        sortedOptionPoints.putAll(optionPoints);
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Perhaps the sorting (lines 403-410) could be performed in the `generateOptionPointsMapping` method so that both the html and csv statistics will be sorted :)

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 2]
```diff
@@ -398,7 +400,16 @@ public String getQuestionResultStatisticsHtml(
 
         DecimalFormat df = new DecimalFormat("#.##");
         
-        for (Entry<String, List<Integer>> entry : optionPoints.entrySet()) {
+        SortedMap<String, List<Integer>> sortedOptionPoints = new TreeMap<String, List<Integer>>(new Comparator<String>() {
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: We can leave it as `Map<String, List<Integer>>` instead of `SortedMap<String, List<Integer>>`. In addition, there's no need for the `Comparator` here, the `TreeMap` will automatically sort in ascending order of the `String` keys

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java (hunk 1 above)
Comment (swxsw) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Noted, will work on it:)

**Comment by swxsw** [comment]
I think it should be done, do verify my changes to test files and expected outputs. Not very certain of it. Had to use comparator for treeMap because if I left it at default the list is displayed in a descending order instead of ascending order. Happy New Year :)

**Comment by JosephineKwa** [comment]
Happy new year to you too :) Its a little late now so I'll take a look at it tomorrow

**Comment by wkurniawan07** [comment]
@swxsw Not going to review this, but you should revert your last three commits since they are problems described in #6287.
In general, if you think that there are failures not caused by your changes, you should not hesitate to clarify with the dev team members.

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 3]
```diff
@@ -475,12 +484,19 @@ public String getQuestionResultStatisticsCsv(
...
     private Map<String, List<Integer>> generateOptionPointsMapping(
             List<FeedbackResponseAttributes> responses) {
         
-        Map<String, List<Integer>> optionPoints = new HashMap<String, List<Integer>>();
+        Map<String, List<Integer>> optionPoints = new TreeMap<String, List<Integer>>(new Comparator<String>() {
+            @Override
+            public int compare(String s1, String s2) {
+                return s2.compareTo(s1);
+            }
+        });
```
Comment (swxsw) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: @JosephineKwa No worries:) But please advise, is it right for me to revert to my previous implementation? Just that the downloaded cvs also has to be sorted?

**Comment by swxsw** [comment]
@wkurniawan07 Noted, will revert

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java (hunk 3 above)
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: By previous implementation do you mean performing the options sort for the html and csv statistics separately, within their respective methods? Yup, it should be alright to do so as long as you are able to sort the options correctly 👍

**Comment by swxsw** [comment]
Yup, I do mean that:) Thanks for the quick reply

**Comment by swxsw** [comment]
@JosephineKwa Should be done now I think, sorted both the display and download file in their individual methods

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 4]
```diff
@@ -397,16 +398,46 @@ public String getQuestionResultStatisticsHtml(
...
         
         for (Entry<String, List<Integer>> entry : optionPoints.entrySet()) {
             
+            if (distributeToRecipients) {
+                
+                String participantIdentifier = entry.getKey();
+                String name = bundle.getNameForEmail(participantIdentifier);
+                
+                identifierMap.put(name, participantIdentifier);
+                sortedOptionPoints.put(name, entry.getValue());
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Hmm if there are 2 or more recipients with the same name, this would result in problems, although I'm not quite sure the best way to solve it myself. Perhaps one way could be to let the key be a concatenation of the recipient name and email. @kanghj any suggestions on how to overcome this?

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 5]
```diff
@@ -397,16 +398,46 @@ public String getQuestionResultStatisticsHtml(
...
+                
+            } else {
+                
+                String option = options.get(Integer.parseInt(entry.getKey()));
+                
+                identifierMap.put(option, entry.getKey());
+                sortedOptionPoints.put(option, entry.getValue());
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Again, this would result in problems if there are 2 or more same options. However this is an issue of its own, to be fixed in another PR, as it does not make sense to have 2 or more same options. Users should be prevented from setting questions with 2 or more same options.

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 6]
```diff
@@ -397,16 +398,46 @@ public String getQuestionResultStatisticsHtml(
         Map<String, List<Integer>> optionPoints = generateOptionPointsMapping(responses);
 
         DecimalFormat df = new DecimalFormat("#.##");
+        /*
```
Comment (swxsw) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Done:)

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java (hunk 4 above)
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Concatenating the name and email should work with little modifications to the code already done in this PR, so we could go for that.

Another possible solution is to stick to the old code and let the recipient's identifier still be the keys of the `sortedOptionPoints` map. Before iterating over the keys, we could sort the keys with a custom comparator. In the comparator, we compare the names obtained using `bundle.getNameForEmail` (either directly, or stored in another map) instead of the keys directly.

Comment (swxsw) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Working on it

Comment (swxsw) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: @JosephineKwa @kanghj Should be done, changes made as discussed. Concatenated the name and email when sorting

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 7]
```diff
@@ -398,14 +399,38 @@ public String getQuestionResultStatisticsHtml(
...
+                String nameEmail = name + participantIdentifier;
+                
+                identifierMap.put(nameEmail, participantIdentifier);
+                sortedOptionPoints.put(nameEmail, entry.getValue());
+                
+            } else {
+                
+                String option = options.get(Integer.parseInt(entry.getKey()));
+                
+                identifierMap.put(option, entry.getKey());
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: This line doesn't look necessary as `identifierMap` isn't used when distributing points among options

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 8]
```diff
@@ -450,14 +475,38 @@ public String getQuestionResultStatisticsCsv(
...
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Similarly this line can be removed :)

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 9]
```diff
@@ -450,14 +475,38 @@ public String getQuestionResultStatisticsCsv(
...
+            }
+        }
+        
+        for (Entry<String, List<Integer>> entry : sortedOptionPoints.entrySet()) {
             String option;
             if (distributeToRecipients) {
-                String teamName = bundle.getTeamNameForEmail(entry.getKey());
-                String recipientName = bundle.getNameForEmail(entry.getKey());
+                String teamName = bundle.getTeamNameForEmail(identifierMap.get(entry.getKey()));
+                String recipientName = bundle.getNameForEmail(identifierMap.get(entry.getKey()));
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Simalar to how it was done for the HTML statistics, it would be good to extract `identifierMap.get(entry.getKey())` into `participantIdentifier`

Comment (swxsw) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: @JosephineKwa Everything is done, just that I have to resolve the merge conflict and regenerate the test files because I let it overwrite my expected test files as manually resolving the expected htmls are too time consuming. Will let you know when it is done:)

Comment (swxsw) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: @JosephineKwa Should be done, please verify:)

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 10]
```diff
@@ -398,14 +399,37 @@ public String getQuestionResultStatisticsHtml(
...
+                
+                identifierMap.put(nameEmail, participantIdentifier);
+                sortedOptionPoints.put(nameEmail, entry.getValue());
+                
+            } else {
+                
+                String option = options.get(Integer.parseInt(entry.getKey()));
+                
+                sortedOptionPoints.put(option, entry.getValue());
+            }
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Remove unnecessary blank lines e.g. 409, 416, 418, etc.? You can still use blank lines to divide code blocks.

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 11]
```diff
@@ -450,14 +474,41 @@ public String getQuestionResultStatisticsCsv(
...
+        }
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: This code block looks very similar to another code block above. Any chance of extracting out the similarity to a utility function, perhaps using generic parameters?

Comment (swxsw) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: @damithc By extracting to a utility function, do you mean writing a function in the commons.util files or just a separate method within the same java file? Not quite getting it

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java (hunk 10 above)
Comment (swxsw) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Should be removed:)

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java (hunk 11 above)
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: That depends on the nature of the function extracted. If it is specific to this class only, we should keep it in this class itself. If it is reusable by other classes, we can move it to a more suitable place.

Comment (swxsw) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Extracted 2 common helper methods into the current java file, still rather specific to the current class.

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 12]
```diff
@@ -378,7 +379,27 @@ public String getQuestionAdditionalInfoHtml(int questionNumber,
...
+    }
+    
+    private void sortForOption(Map.Entry<String, List<Integer>> entry,
+            List<String> optionList,
+            Map<String, List<Integer>> sortedStorageMap) {
+        String option = optionList.get(Integer.parseInt(entry.getKey()));
+        
+        sortedStorageMap.put(option, entry.getValue());
+    }
+
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Put the utility methods below the where they are used, not above. We read in top-down fashion.

Comment (swxsw) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Noted:)

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 13]
```diff
@@ -450,14 +463,29 @@ public String getQuestionResultStatisticsCsv(
...
         DecimalFormat df = new DecimalFormat("#.##");
         
+        Map<String, List<Integer>> sortedOptionPoints = new TreeMap<String, List<Integer>>();
+        
+        Map<String, String> identifierMap = new HashMap<String, String>();
+        
         for (Entry<String, List<Integer>> entry : optionPoints.entrySet()) {
+            
+            if (distributeToRecipients) {
+                sortForRecipient(entry, identifierMap, sortedOptionPoints, bundle);
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Since `optionPoints` contains either recipient mappings or option mappings, we could pass `optionPoints` directly into the method (instead of `entry`), and bring the `for` loop into the method. That way the boolean condition is only checked once before sorting and not on every iteration of the loop. Although it was done this way previously, let's fix it here 😛

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 14]
```diff
@@ -450,14 +463,29 @@ public String getQuestionResultStatisticsCsv(
...
+            } else {
+                sortForOption(entry, options, sortedOptionPoints);
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Similarly, pass `optionPoints` into the method and bring the `for` loop into the method

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java (hunk 13 above)
Comment (swxsw) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Done

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 15]
```diff
@@ -470,6 +492,33 @@ public String getQuestionResultStatisticsCsv(
                + ", Average Points" + Const.EOL
                + fragments + Const.EOL;
     }
+    
+    /**
+     * Used as helper method to put recipient entries from an unsorted map to a sorted map
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Follow the header comment convention for methods. `Puts ...`. Also, don't explain who is using it for what; simply explain what it does. Is this method doing any sorting? If not, the method name should not contain `sort`

Comment (swxsw) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Should be done

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 16]
```diff
@@ -470,6 +492,42 @@ public String getQuestionResultStatisticsCsv(
...
+    /**
+     * Puts recipients from an unsorted map to a sorted map
+     * 
+     * @param recipientMapping      Original map containing recipients
+     * @param identifierMap         Helper map to retrieve email from name concatenated with email string
+     * @param sortedOptionPoints    Sorted map to contain recipient info, recipient concatenated with email used as key
+     * @param bundle
+     */
+    private void putRecipientsInSortedMap(Map<String, List<Integer>> recipientMapping, Map<String, String> identifierMap,
+            Map<String, List<Integer>> sortedOptionPoints, FeedbackSessionResultsBundle bundle) {
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Now that the method name has increased in length, the following would be neater:
```
private void putRecipientsInSortedMap(
        Map<String, List<Integer>> recipientMapping, Map<String, String> identifierMap,
        Map<String, List<Integer>> sortedOptionPoints, FeedbackSessionResultsBundle bundle) {
```

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 17]
```diff
@@ -470,6 +492,42 @@ public String getQuestionResultStatisticsCsv(
...
+    }
+    
+    /**
+     * Puts options from an unsorted map to a sorted map
+     * 
+     * @param optionPoints          Original mapping of option points
+     * @param optionList            List of options in question
+     * @param sortedOptionPoints    Sorted map of option points
+     */
+    private void putOptionsInSortedMap(Map<String, List<Integer>> optionPoints, List<String> optionList,
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Similarly for this:
```
private void putOptionsInSortedMap(
        Map<String, List<Integer>> optionPoints, List<String> optionList,
        Map<String, List<Integer>> sortedOptionPoints) {
```

Comment (swxsw) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: @JosephineKwa Changed accordingly:)

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java (hunk 16 above)
Comment (swxsw) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: Done

**Comment by JosephineKwa** [comment]
@damithc pending your review

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 18]
```diff
@@ -470,6 +492,44 @@ public String getQuestionResultStatisticsCsv(
...
                + fragments + Const.EOL;
     }
+    
+    /**
+     * Puts recipients from an unsorted map to a sorted map
+     * 
+     * @param recipientMapping      Original map containing recipients
+     * @param identifierMap         Helper map to retrieve email from name concatenated with email string
+     * @param sortedOptionPoints    Sorted map to contain recipient info, recipient concatenated with email used as key
+     * @param bundle
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: If there's nothing to say about this parameter, omit the entire line

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java [hunk 19]
```diff
@@ -470,6 +492,44 @@ public String getQuestionResultStatisticsCsv(
...
+     */
+    private void putRecipientsInSortedMap(
+            Map<String, List<Integer>> recipientMapping, Map<String, String> identifierMap,
+            Map<String, List<Integer>> sortedOptionPoints, FeedbackSessionResultsBundle bundle) {
+        for (Entry<String, List<Integer>> entry : recipientMapping.entrySet()) {
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumQuestionDetails.java]: As the signature spans three lines, may be add a blank line below it to separate the body from signature?
